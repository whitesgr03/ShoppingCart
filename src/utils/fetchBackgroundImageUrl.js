import { initialStorage } from "../firebase-config";

const storage = initialStorage();

const fetchBackgroundImageUrl = async page => {
	const ref = storage.ref(`images/${page}/background.jpg`);
	return await ref.getDownloadURL();
};

export { fetchBackgroundImageUrl as default };
