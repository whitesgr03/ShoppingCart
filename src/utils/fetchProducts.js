import { initialFirestore } from "../firebase-config";

const firestore = initialFirestore();

const fetchProducts = async () => {
	const productRef = firestore.collection("products");
	const result = await productRef.get();

	return result.empty
		? (() => {
				throw result;
		  })()
		: result.docs.map(item => {
				const { name, price, url } = item.data();
				return {
					id: item.id,
					name,
					url,
					price,
				};
		  });
};

export { fetchProducts as default };
