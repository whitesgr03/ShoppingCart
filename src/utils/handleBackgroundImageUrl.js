import { initialStorage } from "../firebase-config";

const getBackgroundImageUrl = async page => {
	const storage = initialStorage();
	const ref = storage.ref(`images/${page}/background.jpg`);

	try {
		const url = await ref.getDownloadURL();

		return {
			message: "Get background-image url success.",
			success: true,
			data: url,
		};
	} catch (error) {
		return {
			message: error,
			success: false,
		};
	}
};

export { getBackgroundImageUrl as default };
