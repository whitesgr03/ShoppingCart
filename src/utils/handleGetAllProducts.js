import { initialFirestore } from "../firebase-config";

const handleGetAllProducts = async () => {
	const firestore = initialFirestore();
	const productsRef = firestore.collection("products");

	const result = await productsRef.get();

	const allProductsResult = result.empty
		? []
		: result.docs.map(item => ({
				id: item.id,
				...item.data(),
		  }));

	return allProductsResult;
};

export default handleGetAllProducts;
