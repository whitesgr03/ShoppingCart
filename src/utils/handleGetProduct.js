import { initialFirestore } from "../firebase-config";

const handleGetProduct = async productId => {
	const firestore = initialFirestore();
	const productRef = firestore.collection("products").doc(productId);

	const result = await productRef.get();

	const productResult = {
		id: result.id,
		...result.data(),
	};

	return productResult;
};

export default handleGetProduct;
