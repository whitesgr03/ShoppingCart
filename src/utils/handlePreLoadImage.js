const preLoadImage = url =>
	new Promise(resolve => {
		const img = new Image();
		img.addEventListener("load", () => resolve(url), { once: true });
		img.src = url;
	});
	
export default preLoadImage;
