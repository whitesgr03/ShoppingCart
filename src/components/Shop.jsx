import { useState, useEffect } from "react";
import { Outlet, Link, useLocation, useOutletContext } from "react-router-dom";

import Icon from "@mdi/react";
import { mdiMagnify, mdiLoading } from "@mdi/js";

import { fetchResource } from "../utils/utils";

const Shop = () => {
	const { onAddItem, onToggleModal } = useOutletContext();
	const { pathname, key: currentKey } = useLocation();

	const isShopRoute = pathname === "/shop";

	const [products, setProducts] = useState(null);
	const [filterText, setFilterText] = useState("");
	const [previousKey, setPreviousKey] = useState(currentKey);

	useEffect(() => {
		const handleFetch = async () => {
			setProducts(await fetchResource());
		};
		handleFetch();
	}, []);

	useEffect(() => {
		const isKeyChanged = previousKey !== currentKey;

		const handleClearFilterText = () => {
			setFilterText("");
			setPreviousKey(currentKey);
		};

		isShopRoute && isKeyChanged && handleClearFilterText();
	});

	return (
		<div data-testid="shop" className="shop">
			{products ? (
				<>
					<Navbar
						isShopRoute={isShopRoute}
						filterText={filterText}
						onFilterTextChange={setFilterText}
					/>

					<Outlet
						context={{
							products,
							filterText,
							onAddItem,
							onToggleModal,
						}}
					/>
				</>
			) : (
				<div data-testid="loading" className="loading">
					<Icon path={mdiLoading} spin={1} size={3} />
					{"Loading..."}
				</div>
			)}
		</div>
	);
};

const Navbar = ({
	isShopRoute, 
	filterText,
	onFilterTextChange
}) => {
	const [active, setActive] = useState(false);

	const handleSubmit = e => {
		e.preventDefault();
		onFilterTextChange(...new FormData(e.target).values());
	};

	const handleChange = e => {
		onFilterTextChange(e.target.value);
	};

	const handleActiveSearchBarBorder = () => {
		setActive(true);
	};

	const handleDeActivateSearchBarBorder = () => {
		setActive(false);
	};

	return (
		<div className="navigation" data-testid="navigation">
			<ul className="category">
				<li>
					<Link to="/shop">Shop</Link>
				</li>
				<li> / </li>
				<li>
					<Link to="/shop">Category</Link>
				</li>
			</ul>
			{isShopRoute && (
				<form
					data-testid="searchBar"
					className={`searchBar ${active ? "active-border" : ""}`}
					onFocus={handleActiveSearchBarBorder}
					onBlur={handleDeActivateSearchBarBorder}
					onSubmit={handleSubmit}
				>
					<div className="search-border">
						<input
							data-testid="search"
							type="search"
							name="search"
							placeholder="search..."
							value={filterText}
							onChange={handleChange}
						/>
					</div>
					<button type="submit">
						<Icon path={mdiMagnify} />
					</button>
				</form>
			)}
		</div>
	);
};

const Products = () => {
	const { products, filterText } = useOutletContext();

	const filterProducts =
		filterText === ""
			? products
			: products.filter(
					item =>
						item.name
							.toLowerCase()
							.indexOf(filterText.toLowerCase().trim()) !== -1
			  );

	const productList =
		products.length > 0 ? (
			filterProducts.map(product => (
				<div className="item" key={product.id}>
					<Link
						className="image-link"
						to={`/shop/${product.id}`}
						state={{ product }}
					>
						<img src={product.url} alt={product.name} />
					</Link>
					<div className="info">
						<div className="title">
							<Link
								className="text-link"
								to={`/shop/${product.id}`}
								state={{ product }}
							>
								{product.name}
							</Link>
						</div>
						<div className="price">${product.price.toFixed(2)}</div>
					</div>
				</div>
			))
		) : (
			<div>{"loading..."}</div>
		);

	return (
		<div
			data-testid="products"
			className={`products ${products.length > 0 ? "" : "loading"}`}
		>
			{productList}
		</div>
	);
};

const ProductInfo = () => {
	const { state } = useLocation();

	const { onAddItem, onToggleModal } = useOutletContext();

	const maxOptions = 10;

	const options = Array.from({ length: maxOptions })
		.fill([])
		.map((_, i) => (
			<option key={i} value={i + 1}>
				{i + 1}
			</option>
		));

	const handleSubmit = e => {
		e.preventDefault();

		const formData = new FormData(e.target).entries();

		const entriesNumeric = [...formData].map(([key, value]) =>
			key === "quantity" ? [key, Number(value)] : [key, value]
		);

		const newProduct = {
			...state.product,
			...Object.fromEntries(entriesNumeric),
		};

		onAddItem(newProduct);
		onToggleModal("showAlert");
	};

	return state ? (
		<div className="productInfo">
			<div className="image">
				<img src={state.product.url} alt={state.product.name} />
			</div>
			<div className="info">
				<div className="description">
					<h2 className="title">{state.product.name}</h2>
					<p>
						Lorem ipsum dolor sit amet consectetur adipisicing elit.
						Magnam itaque nisi ipsa pariatur perspiciatis nemo.
					</p>
					<p className="price">${state.product.price.toFixed(2)}</p>
				</div>

				<form onSubmit={handleSubmit}>
					<label htmlFor="quantity">
						Quantity
						<select name="quantity" id="quantity" defaultValue={1}>
							{options}
						</select>
					</label>
					<button className="slide" type="submit">
						Add to Cart
					</button>
				</form>
			</div>
		</div>
	) : (
		<div data-testid="productError" className="productError">
			<div className="error">
				<p>Our apologies, there has been an error.</p>
				<p>Please come back later or return to the previous page.</p>
			</div>
		</div>
	);
};

export { Shop as default, Navbar, Products, ProductInfo };
