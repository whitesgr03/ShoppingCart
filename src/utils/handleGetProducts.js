import { initialFirestore } from "../firebase-config";
import handlePreLoadImage from "./handlePreLoadImage";

const handleGetProducts = async () => {
	const firestore = initialFirestore();
	const productRef = firestore.collection("products");

	const result = await productRef.get();

	await Promise.all(
		result.docs.map(item => {
			const obj = { ...item.data() };
			return handlePreLoadImage(obj.url);
		})
	);

	const productsResult = result.docs.map(item => ({
		id: item.id,
		...item.data(),
	}));

	return productsResult;
};

export default handleGetProducts;
