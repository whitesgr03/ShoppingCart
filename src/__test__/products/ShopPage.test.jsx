import { render, screen, waitFor } from "@testing-library/react";
import { RouterProvider, createMemoryRouter } from "react-router-dom";

import Shop from "../../components/products/ShopPage";

import { fetchResource } from "../../utils/utils";

const mockFetchData = [
	{
		id: 0,
		name: "fakeBag",
		url: "../",
		price: 19.9,
	},
	{
		id: 1,
		name: "fakePants",
		url: "../",
		price: 19.9,
	},
];

jest.mock("../../utils/utils", () => ({
	...jest.requireActual("../../utils/utils"),
	fetchResource: jest.fn(() => mockFetchData),
}));

describe("Renders Shop Component", () => {
	it("Should return Shop DOM", async () => {
		const routes = [
			{
				path: "/",
				element: <Shop />,
			},
		];

		const router = createMemoryRouter(routes, {
			initialEntries: ["/"],
		});

		render(<RouterProvider router={router} />);

		const loading = screen.getByTestId("loading");

		expect(loading).toHaveClass("loading");

		expect(fetchResource).toBeCalledTimes(1);

		await waitFor(() => {
			expect(loading).not.toBeInTheDocument();
		});
	});
});
