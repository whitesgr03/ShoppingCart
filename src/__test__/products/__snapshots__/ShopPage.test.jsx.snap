// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Shop Component Should return Shop DOM 1`] = `
<div>
  <div
    class="shop"
  >
    <div
      class="navigation"
    >
      <ul
        class="category"
      >
        <li>
          <a
            href="/shop"
          >
            Shop
          </a>
        </li>
        <li>
           / 
        </li>
        <li>
          <a
            href="/shop"
          >
            Category
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
